### ISSUE TITLE
Export YOLO11-pose to onnx with NMS

### ISSUE THREAD

**Issue by russel0719** [ultralytics/ultralytics#17149]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, Thank you for your great works!

I want to make onnx exported model of YOLO11-pose. but .export function of YOLO didn't support NMS.
So I want to make custom wrapper class to export yolo11-pose to onnx with NMS.
There are also many similar issues with this question, but i can't understand some
Below is my custom wrapper class implementation

```python
class Wrapper(nn.Module):
    def __init__(self, model_path):
        super().__init__()
        self.model = YOLO(model=model_path).model


    def forward(self, x):
        pred = self.model(x)[0]      # [1, 49, 21504] my model pose object (class, x, y, w, h, x1, y1, ..., x22, y22)

        pred = pred.permute(0, 2, 1)
        
        boxes, scores, keypoints = pred.split([4, 1, 44], dim=-1)
        boxes = torchvision.ops.box_convert(boxes, in_fmt='cxcywh', out_fmt='xyxy')  # [1, 21504, 4]
        boxes = boxes.squeeze(0)
        scores = scores.squeeze(0).squeeze(-1)
        keypoints = keypoints.squeeze(0).reshape(-1, 22, 2)

        keep = torchvision.ops.nms(boxes, scores, iou_threshold=0.45)

        boxes = boxes[keep]
        scores = scores[keep]
        keypoints = keypoints[keep]

        idx = np.argmax(scores.cpu().detach().numpy())
        return boxes[idx], keypoints[idx]
```

and my export code is

```python
input_np = np.random.randn(1, 3, 1024, 1024).astype(np.float32)
input_tensor = torch.from_numpy(input_np).float() / 255.0
wrapped_model = Wrapper(model_path=model_path)
torch.onnx.export(
    wrapped_model,                      # Wrapped YOLO pose model with NMS
    input_tensor,                       # Example input tensor
    output_path,                        # Path to save ONNX model
    export_params=True,                 # Store the trained parameter weights inside the model file
    opset_version=11,                   # ONNX opset version
    do_constant_folding=True,           # Simplify the model
    input_names=['input'],              # Input name in ONNX model
    output_names=['boxes', 'keypoints'],  # Output names
    dynamic_axes={'input': {0: 'batch_size'},   # Dynamic batch size
                'boxes': {0: 'batch_size'},
                'keypoints': {0: 'batch_size'}}
    )
```

my questions are
1. with these class and code, model is successfully exported but when i inference with this exported model, outputs are different from original model's output. do i use right code? can you check it?
2. if there are no problem with these codes, how can i use exported model? can you give me some example code?

Thank you for your help...

**Comment by UltralyticsAssistant** [comment]
👋 Hello @russel0719, thank you for reaching out and for your interest in Ultralytics 🚀! This is an automated response, but rest assured, an Ultralytics engineer will assist you shortly.

We recommend taking a look at our [Docs](https://docs.ultralytics.com), where you can find comprehensive guides and examples, including many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples.

If this is a 🐛 Bug Report, it would be helpful to have a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to facilitate debugging. Please include any logs if available.

Your question seems to relate to custom modifications and model export. If so, please ensure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) and share more specific details on your training setup when possible.

To further engage with others, consider joining our community:
- Join real-time discussions on [Discord](https://ultralytics.com/discord) 🎧.
- For detailed discussions, check out [Discourse](https://community.ultralytics.com).
- Share insights on our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

To ensure compatibility, please upgrade to the latest version of the `ultralytics` package, and verify all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are met in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Consider running YOLO in one of these up-to-date verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If the CI badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, ensuring correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across platforms.

**Comment by Y-T-G** [comment]
Does it work before conversion?

Use torch instead of numpy and don't move to CPU or convert to numpy

`idx = np.argmax(scores.cpu().detach().numpy())`

You're not applying any confidence threshold before NMS.

**Comment by russel0719** [comment]
Hi @Y-T-G 
Thank you for your reply!

I only tested with `model.predict(source='image_path')` like this.

```python
class Wrapper(nn.Module):
    def __init__(self, model_path):
        super().__init__()
        self.model = YOLO(model=model_path).model


    def forward(self, x):
        pred = self.model(x)[0]      # [1, 49, 21504]

        pred = pred.permute(0, 2, 1)
        
        boxes, scores, keypoints = pred.split([4, 1, 44], dim=-1)
        
        score_idx = scores.squeeze(-1) > 0.25
        boxes = boxes[score_idx]
        keypoints = keypoints[score_idx]
        scores = scores[score_idx]

        boxes = torchvision.ops.box_convert(boxes, in_fmt='cxcywh', out_fmt='xyxy')  # [21504, 4]
        scores = scores.squeeze(-1)
        keypoints = keypoints.reshape(-1, 2, 22).permute(0, 2, 1)

        keep = torchvision.ops.nms(boxes, scores, iou_threshold=0.7)

        boxes = boxes[keep]
        scores = scores[keep]
        keypoints = keypoints[keep]

        idx = torch.argmax(scores)
        return boxes[idx], keypoints[idx]
```

i tried this code, but i got keypoints with value less than 0 like below
[     125.95      177.91] [     -129.2      11.668]
[     125.95      177.91] [     187.06      146.27]
[     -129.2      11.668] [    -120.32      18.539]
[     187.06      146.27] [    -120.32      18.539]
[      130.8       186.3] [    -79.257      27.442]
[      130.8       186.3] [     189.81      129.18]
[    -79.257      27.442] [    -93.899      45.423]
[     189.81      129.18] [    -93.899      45.423]
[     142.17      183.09] [    -78.139      47.764]
[     142.17      183.09] [     203.88      134.37]
[    -78.139      47.764] [    -86.118      42.796]
[     203.88      134.37] [    -86.118      42.796]
[     132.06      179.73] [    -43.627      55.156]
[     132.06      179.73] [     190.18      135.78]
[    -43.627      55.156] [    -34.981      90.433]
[     190.18      135.78] [    -34.981      90.433]
[     144.45      185.06] [    -33.303       75.57]
[     144.45      185.06] [     201.64      137.27]
[    -33.303       75.57] [    -16.617      67.269]
[     201.64      137.27] [    -16.617      67.269]
[     120.84      173.48] [     8.9048      52.796]

box value also contain negative value
[      114.9     -10.958      200.48      69.145]

**Comment by russel0719** [comment]
@Y-T-G Oh, I found my mistakes. Sorry for the inconvenience.
I fixed the codes and got good results.

Thank you!!

**Comment by Y-T-G** [comment]
I created this with proper batching and fixed output size:

```python
from ultralytics import YOLO
from torch import nn
import torch
import torchvision
from ultralytics.utils.ops import xywh2xyxy
import cv2

img = cv2.imread("ultralytics/assets/bus.jpg")

model = YOLO("yolov8n-pose.pt")

_ = model()

prep = model.predictor.preprocess([img, img.transpose(1,0,2)])

class Model(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model.model
    def forward(self, x):
        pred = self.model(x)[0]
        pred = pred.permute(0, 2, 1)
        boxes, scores, keypoints = pred.split([4, 1, 51], dim=2)
        scores, classes = scores.max(dim=-1)
        # Only send top 300 to NMS
        scores, keep = torch.topk(scores, 300)
        boxes, keypoints = [i.gather(1, keep.unsqueeze(-1).repeat(1, 1, i.shape[-1])) for i in (boxes, keypoints)]
        classes = classes.gather(1, keep)[..., None]
        scores = scores[..., None]
        boxes = xywh2xyxy(boxes)
        # Fixed output size: 100
        out = torch.zeros(boxes.shape[0], 100, boxes.shape[-1] + classes.shape[-1] + scores.shape[-1] + keypoints.shape[-1])
        for i in range(boxes.shape[0]):
            box, cls, score, kp = boxes[i], classes[i], scores[i], keypoints[i]
            box = box + cls[0] # Class specific NMS
            keep = torchvision.ops.nms(box, score.squeeze(), 0.45)
            out[i, :keep.shape[0]] = torch.cat([box[keep], cls[keep], score[keep], kp[keep]], dim=1)
        # Output shape: [num_images, 4 box corrdinates + 1 class index + 1 score value + 17 x 3 keypoint coordinates]
        return out

onnxm = Model(model)

torch.onnx.export(
    onnxm,
    prep,
    "onnx_nms.onnx",
    export_params=True,                 # Store the trained parameter weights inside the model file
    opset_version=11,                   # ONNX opset version
    do_constant_folding=True,           # Simplify the model
    input_names=['input'],              # Input name in ONNX model
    output_names=['outputs'],  # Output names
    dynamic_axes={'input': {0: 'batch_size'},   # Dynamic batch size
                'boxes': {0: 'batch_size'},
                'keypoints': {0: 'batch_size'}}
     )
```

**Comment by russel0719** [comment]
@Y-T-G Sorry for late reply!

I've never thought about topk method in NMS. Thank you for your good idea.
I will apply your idea.

Thank you!!

**Comment by UltralyticsAssistant** [comment]
The associated pull request [#18484](https://github.com/ultralytics/ultralytics/pull/18484) has been merged, and it includes a potential fix that may resolve this issue! 🎉  

### Relevant Key Changes from the PR:
This update introduces extended export functionalities and post-processing improvements for object detection models, addressing several limitations in the workflow. Notable changes include:
- **Non-Maximum Suppression (NMS) Integration**: Exported models (TorchScript, ONNX, OpenVINO, TensorRT, TensorFlow) now include embedded NMS, providing seamless post-processing capabilities.
- **Enhanced Post-Processing Logic**: Predictions for various tasks (detection, segmentation, pose, and OBB) have been refined, including additional attribute support, like `keypoints`, for improved handling of prediction results.
- **Flexibility in Export Configurations**: Users can now toggle NMS during the export process, enabling easier deployment customization.
- **Better Metadata Handling**: Updates for improved export compatibility across formats.
- 🆙 Version upgraded to `8.3.67` for tracking this new functionality.

These changes aim to streamline the export process while improving runtime efficiency and usability for deployment scenarios.  

### Contributors ❤️
This valuable update was made possible due to contributions from [Y-T-G](https://github.com/Y-T-G), [lifafang](https://github.com/lifafang), [boorad](https://github.com/boorad), [glenn-jocher](https://github.com/glenn-jocher), [Laughing-q](https://github.com/Laughing-q), and [inisis](https://github.com/inisis). A big thank you to everyone involved! 🙌  

### How to Test the Changes:
To verify if this resolves your issue:
- Install the latest changes directly from the `main` branch:
  ```bash
  pip install git+https://github.com/ultralytics/ultralytics.git@main
  ```
- Or you can wait for the next official release on PyPI if preferred.

Let us know if this resolves the issue or if you encounter any remaining problems. Your feedback is invaluable in ensuring the robustness of these updates.  

Thanks again for bringing up this issue! 🙏 If anything persists or additional adjustments are needed, feel free to share more details.
